Monitor-move lock race test is flaky under a constrained scheduler

Follow-up from the #1224 campaign. Introduced by #1238.

## Neru version

`dev` — source build of `main` at `06414b30` (`refactor(modes): let each mode put itself back on a display that changed (#1241)`). Not a released-version bug; it is a test defect on `main`.

## Operating system

macOS

## OS version

macOS 26.5.2 (build 25F84), arm64, Go 1.26.4. Not platform-specific — the test is pure Go with no build tags.

## Navigation mode

n.a. — unit test in `internal/app/modes`.

## What happened

`TestRefreshActiveModeOnNewScreen_ReadsTheModeSessionUnderTheHandlerLock` fails roughly half the time when the scheduler is constrained:

```
--- FAIL: TestRefreshActiveModeOnNewScreen_ReadsTheModeSessionUnderTheHandlerLock (0.00s)
    monitor_test.go:119: no monitor-move refresh reached the hints session, so nothing raced
```

Measured on `main` at `06414b30`: **10 failures in 20 runs** at `GOMAXPROCS=1`. It passes reliably at default `GOMAXPROCS`, which is why five PRs (#1238–#1241) went green over it.

The failure is scheduler-dependent, not a real race: the test starts concurrent work and then asserts that the refresh actually reached the hints session, but nothing makes that guaranteed to happen before the assertion. When the scheduler does not cooperate, the test correctly reports that it observed nothing — and fails.

## What did you expect

The test either observes the interleaving it is written to observe, or is structured so that not observing it is not a failure. A test that asserts on a scheduler-decided count is not a test of the lock.

## Steps to reproduce

1. `git checkout main` (at `06414b30` or later)
2. `GOMAXPROCS=1 go test -race -run 'TestRefreshActiveModeOnNewScreen_ReadsTheModeSessionUnderTheHandlerLock' -count=20 ./internal/app/modes/`
3. Observe ~50% failures with the message above.

At default `GOMAXPROCS` the same command passes, so CI has not caught it.

## Config (relevant sections)

n.a.

## Additional context

**The fix is already known and used twice in this repo.** Both #1240 and #1241 hit exactly this class of defect in their own new race tests, and `deadlock-reviewer` flagged both. The pattern applied in each case: prove deterministically that one refresh reaches the session *before* starting the concurrent loops, then only log the concurrent count rather than asserting on it. #1240 went from 3/40 failures to 0 at `-count=40` and at `GOMAXPROCS=1 -count=20`; #1241 went from 7/12 failures at `-cpu=1` to 13/13 green. This test needs the same treatment.

Failing in this direction is the safe one — the test reports that it could not observe the path rather than passing falsely — but it will redden CI on a loaded or constrained runner, and a test that fails for reasons unrelated to the behaviour it guards trains people to re-run rather than read.

Note the test name still says `RefreshActiveModeOnNewScreen`; the function it covers was renamed to `refreshActiveModeForMonitorMove` in #1238. Worth renaming in the same change.

--- a/docs/adr/0004-mode-lifecycle-dispatch-under-one-lock-hold.md
+++ b/docs/adr/0004-mode-lifecycle-dispatch-under-one-lock-hold.md
@@ -85,7 +85,7 @@ of keeping mode knowledge in a switch the modes work exists to delete.
   under the same lock. What it buys is the end of a real data race — the
   session's filter roles, strategy and label-direction overrides were read with
   no lock held while a concurrent mode exit reset them. That is pinned by
-  `TestRefreshActiveModeOnNewScreen_ReadsTheModeSessionUnderTheHandlerLock`,
+  `TestRefreshActiveModeForMonitorMove_ReadsTheModeSessionUnderTheHandlerLock`,
   which fails under `-race` against the dispatch this ADR replaces.
 - **The hold is now as long as the slowest refresh**, so that refresh had to
   be given a bound. The hints walk gets the same `HintTimeout` budget the

--- a/internal/app/modes/monitor_test.go
+++ b/internal/app/modes/monitor_test.go
@@ -23,6 +23,10 @@ import (
 // tests. Its contents do not matter; only that it is not where the mode was.
 var targetDisplay = image.Rect(1920, 0, 3840, 1080)
 
+// monitorMoveFilterRole is a role a hint session was activated with: exiting
+// the mode resets it, which is the state a concurrent refresh reads.
+const monitorMoveFilterRole = "button"
+
 // newMonitorMoveHintsHandler builds a handler with hints mode active and
 // enough wiring for the monitor-move refresh to run end to end: a hint service
 // to regenerate against, a hints context holding the session's flags, and the
@@ -59,15 +63,15 @@ func newMonitorMoveHintsHandler(regenerations *atomic.Int64) *Handler {
 	})
 
 	handler.appState.SetMode(domain.ModeHints)
-	handler.hints.Context.SetFilterRoles([]string{"button"})
+	handler.hints.Context.SetFilterRoles([]string{monitorMoveFilterRole})
 
 	return handler
 }
 
-// TestRefreshActiveModeOnNewScreen_ReadsTheModeSessionUnderTheHandlerLock is
-// the concurrency test the modes area guide asks of a locking change: it must
-// fail under -race if the monitor-move dispatch reads a mode's session state
-// outside h.mu.
+// TestRefreshActiveModeForMonitorMove_ReadsTheModeSessionUnderTheHandlerLock
+// is the concurrency test the modes area guide asks of a locking change: it
+// must fail under -race if the monitor-move dispatch reads a mode's session
+// state outside h.mu.
 //
 // A monitor move races a mode exit by construction — the warp is animated and
 // the user can press escape during it — and the two touch the same fields.
@@ -78,33 +82,66 @@ func newMonitorMoveHintsHandler(regenerations *atomic.Int64) *Handler {
 // was taken and the race detector saw it. Dispatching under one hold is what
 // closes it, so this test is the guard on that hold rather than on the
 // refresh's result.
-func TestRefreshActiveModeOnNewScreen_ReadsTheModeSessionUnderTheHandlerLock(t *testing.T) {
+//
+// How often the two loops below actually meet is the scheduler's business, so
+// the refresh is proven to reach a session up front rather than by asserting on
+// a count the scheduler decides.
+func TestRefreshActiveModeForMonitorMove_ReadsTheModeSessionUnderTheHandlerLock(t *testing.T) {
 	var regenerations atomic.Int64
 
 	handler := newMonitorMoveHintsHandler(&regenerations)
 
+	// A refresh reaches the session at all — asserted on its own, before the two
+	// loops below, because whether either of them ever catches the other
+	// mid-round is up to the scheduler. Without this the whole test could pass
+	// on a dispatch that never reached a session.
+	handler.refreshActiveModeForMonitorMove(context.Background(), targetDisplay)
+
+	if regenerations.Load() == 0 {
+		t.Fatal("a monitor-move refresh did not reach an open hints session")
+	}
+
+	regenerations.Store(0)
+
 	const rounds = 200
 
 	start := make(chan struct{})
+	done := make(chan struct{})
 
 	var waitGroup sync.WaitGroup
 
 	waitGroup.Go(func() {
 		<-start
 
+		defer close(done)
+
 		for range rounds {
-			// Re-enter hints so the next exit has a session to tear down,
-			// the way a repeating activation does. Exiting is the locked
-			// writer: it resets the session flags the refresh reads.
+			// Re-enter hints so the next exit has a session to tear down, the
+			// way a repeating activation does. Both sides of the round are
+			// locked writers — an activation and an exit — and the exit resets
+			// the session flags the refresh reads.
+			handler.mu.Lock()
 			handler.appState.SetMode(domain.ModeHints)
+			handler.hints.Context.SetFilterRoles([]string{monitorMoveFilterRole})
+			handler.mu.Unlock()
+
 			handler.ExitMode()
 		}
 	})
 
+	// The refresher runs flat out for as long as the mode is being entered and
+	// exited, rather than for a round count of its own: under the race detector
+	// the exit side is much the slower of the two.
 	waitGroup.Go(func() {
 		<-start
 
-		for range rounds {
+		for {
+			select {
+			case <-done:
+				return
+			default:
+			}
+
 			handler.refreshActiveModeForMonitorMove(context.Background(), targetDisplay)
 		}
 	})
@@ -112,22 +149,23 @@ func TestRefreshActiveModeOnNewScreen_ReadsTheModeSessionUnderTheHandlerLock(t *
 	close(start)
 	waitGroup.Wait()
 
-	// Without this the test could pass by never refreshing anything: the
-	// window it guards only exists while a refresh is reading the session a
-	// concurrent exit is clearing.
-	if regenerations.Load() == 0 {
-		t.Fatal("no monitor-move refresh reached the hints session, so nothing raced")
-	}
+	// Reported rather than asserted: the exit side re-locks the moment it
+	// unlocks, so how many of the concurrent refreshes find a session open is
+	// the scheduler's decision. Asserting on it makes the test flaky rather
+	// than strict — the guarantee that a refresh reaches a session at all is
+	// the one made before the loops started.
+	t.Logf("%d of the concurrent refreshes reached an open session", regenerations.Load())
 
 	if got := handler.appState.CurrentMode(); got != domain.ModeIdle {
 		t.Fatalf("mode after the last exit = %v, want idle", got)
 	}
 }
 
-// TestRefreshActiveModeOnNewScreen_LeavesTheOverlayAloneWhileIdle pins the arm
-// the mode map answers by having no entry: a move with nothing open must not
-// bring a frame back up, which is what the deleted idle switch arm used to say.
-func TestRefreshActiveModeOnNewScreen_LeavesTheOverlayAloneWhileIdle(t *testing.T) {
+// TestRefreshActiveModeForMonitorMove_LeavesTheOverlayAloneWhileIdle pins the
+// arm the mode map answers by having no entry: a move with nothing open must
+// not bring a frame back up, which is what the deleted idle switch arm used to
+// say.
+func TestRefreshActiveModeForMonitorMove_LeavesTheOverlayAloneWhileIdle(t *testing.T) {
 	overlay := &portmocks.MockOverlayPort{}
 
 	handler := newHandlerWithState(handlerState{

